### ISSUE TITLE
resolved SetLinkDNS() does not update generated resolv.conf

### ISSUE THREAD

**Issue by martinpitt** [systemd/systemd#3225]
### Submission type
- [X] Bug report
- [ ] Request for enhancement (RFE)

_NOTE: Do not submit anything other than bug reports or RFEs via the issue tracker!_
### systemd version the issue has been seen with

> 229, upstream master

_NOTE: Do not submit bug reports about anything but the two most recently released systemd versions upstream!_
### Used distribution

> Ubuntu
### In case of bug report: Steps to reproduce the problem

I want to unify DNS handling to resolved. But on desktops we can't use networkd for everything yet, but need to use NetworkManager (wifi, 3G), and legacy setups that call "dhclient" directly should also work, i. e. pick up DNS servers from DHCP that way. The fallback mode of keeping `/etc/resolv.conf` as a plain file works fine in general, dhclient pushes into that and resolved reads it. But as this file potentially has multiple writers, I'd like to retain the multiplexing feature of resolvconf or resolved, i. e. for me `/etc/resolv.conf` is a symlink to `/run/systemd/resolve/resolv.conf` as recommended in https://wiki.freedesktop.org/www/Software/systemd/writing-network-configuration-managers/ .

I have an ethernet interface (ens3) with ifindex==2 which is not managed by networkd, but (indirectly) uses `dhclient`. Initially resolved just uses the fallback Google name servers and resolved has no per-link DNS servers:

```
# cat /run/systemd/resolve/resolv.conf
# This file is managed by systemd-resolved(8). Do not edit.
[...]
nameserver 8.8.8.8

# busctl call org.freedesktop.resolve1 /org/freedesktop/resolve1 org.freedesktop.resolve1.Manager GetLink i 2
o "/org/freedesktop/resolve1/link/_32"
# busctl introspect org.freedesktop.resolve1 /org/freedesktop/resolve1/link/_32  | grep '\.DNS '
.DNS                                property  a(iay)    0            -
```

Now I try to push the new nameserver, as per above wiki page:

```
# busctl call org.freedesktop.resolve1 /org/freedesktop/resolve1 org.freedesktop.resolve1.Manager SetLinkDNS 'ia(iay)' 2 1 2 4 10 0 2 3

# busctl introspect org.freedesktop.resolve1 /org/freedesktop/resolve1/link/_32  | grep '\.DNS '
.DNS                                property  a(iay)    1 2 4 10 0 2 3 -
```

(Note that there is no journal log entry at all about this, not even with debug).
This actually does cause resolved to use the new name, which can be demonstrated by setting an invalid name server and then DNS resolution takes 30 seconds (until it falls back to `dns` after `resolve`).

But setting new name servers with `SetLinkDNS()` does not cause the generated `resolv.conf` to get updated, it still has the default Google name servers, so that DNS resolution that does not go via resolved, or more importantly, admins inspecting `resolv.conf` to see which servers actually get used, see the wrong ones.

Arguably this is sort of half-correct: the DNS server only applies to the given link, not globally. But resolved does not have an API (D-Bus or files in `/run`, which I'd actually prefer) to set a "global" name server, and practically, every DHCP response comes from a particular link but is then used in some "global" sense.

Also, when `resolved` picks up configuration from `networkd`, i. e. when I _do_ use networkd to manage `ens3`, then information gets propagated via `/run/systemd/netif/links/2` → `/run/systemd/netif/state` (which has `DNS=10.0.2.3`) → resolved (via inotify/network monitor) → `resolv.conf`, i. e. network _does_ treat the picked up DNS servers (from DHCP) as "global" and resolved puts it into the generated `resolv.conf`.

So in summary, resolved should either update `resolv.conf` on `SetLinkDNS()` or grow a file or D-Bus interface to set global name servers, to get an usable/inspectable `resolv.conf`.

**Comment by martinpitt** [comment]
FWIW, this is not a regression (it's a new API in 229) and not blocking me personally, that's why I didn't originally milestone it to v230. I'm also not quite clear yet _how_ it should be fixed, i. e. how it's supposed to behave.
